### ISSUE TITLE
iTunes ("Music" now?) Plug-in "Search by Artist" trigger doesn't work in macOS Catalina

### ISSUE THREAD

**Issue by taberrr** [quicksilver/Quicksilver#2495]
For questions or tech support, please start with the [users’ group][1]. If you know the problem is in Quicksilver itself, report it here.

[1]: https://groups.google.com/forum/#!forum/blacktree-quicksilver

**Describe the bug**
I assigned a trigger to "Search Artists" for the iTunes plug-in but when I press the keys, nothing happens. It's probably a bit early to report, but I'm using macOS Catalina Beta 1. I've added Quicksilver to "Accessibility" and "Input Monitoring" in the Security & Privacy prefs pane. Also, when I try going to "Run" in QS and type in "search by" I get the attached result ("Search Contents of (null)")

**To Reproduce**
1. Install the iTunes Plug-in
2. Assign a trigger to "Search Artists", eg: Cmd+Shift+Ctrl+A
3. Press trigger keys, nothing happens
4. Tried rebooting and relaunching QS, no love :(

**Expected behavior**
Quicksilver dialogue opens with "Search Artist" icon showing

**Screenshots**
<img width="365" alt="Screen Shot 2019-06-16 at 4 08 49 AM" src="https://user-images.githubusercontent.com/95786/59561554-dbda8e00-8fef-11e9-8939-604bcd760d7b.png">

**What version of macOS?**
10.15 Beta (19A471t)

**What version of Quicksilver?**
1.6.0 (4025)

**Comment by skurfer** [comment]
Yeah, the current iTunes plug-in isn’t going to work with Music. I think most of the scripting stuff is the same, but the name and bundle ID of the app are different, so the plug-in is talking to something that isn’t there.

The biggest difficulty will be that they removed the XML version of your library that QS was reading from. I assume (hope) iTunes DB is still available in Catalina so we’ll have some way to read the library.

**Comment by taberrr** [comment]
So back in June I uninstalled Catalina and went back to Mojave because of this, haha. I use this shortcut literally all day long, every day, while I work. It's the only way I listen to music!

Now that Catalina is officially out, I re-installed it, and I miss this QS plugin so bad!! :( @skurfer I see you self-assigned this issue in Sept, which is awesome and gives me hope.  👍

I assume parsing the iTunes DB is no easy task, so best of luck there! If you need any help testing dev builds or anything, lemme know! Or, happy to buy you a beer or 2 to boost morale. :D

**Comment by skurfer** [comment]
> I use this shortcut literally all day long, every day, while I work. It's the only way I listen to music!

Same. Although they brought back the Column Browser in 10.15.2, so at least there’s a way to muddle through now.

> Now that Catalina is officially out, I re-installed it, and I miss this QS plugin so bad!! :(

Right? And what was the upside?

> I assume parsing the iTunes DB is no easy task, so best of luck there!

Well, there’s an official API that doesn’t look too bad, if you can get it to work. It appears to be quite picky about only allowing interactions from a signed app, and I’m having trouble with code signing at the moment, so I can’t even really _start_ the conversion. Other than that… 🙄

> If you need any help testing dev builds or anything, lemme know! Or, happy to buy you a beer or 2 to boost morale. :D

I have plenty of beer, thanks. Got any **free time** I can use?

**Comment by j256** [comment]
Any work on this?   I'm surprised that a new plugin hasn't been created.  Is it really as each as tweaking the bundle-id and application name?

**Comment by skurfer** [comment]
> Is it really as each as tweaking the bundle-id and application name?

No, as I mentioned above, the XML file we used to get library information is gone. We'll need to rewrite everything to use the iTunes DB API.

**Comment by mrstanwell** [comment]
@skurfer I do devops work and have some experience signing and notarizing a macOS app for non-app-store distribution.  I don't do much actual dev work in Xcode, and I can't claim to fully understand the entire signing/notarization process, but I _have_ shipped a signed application that runs successfully on Catalina.  Can I help?  I've relied on QS probably for well over a decade, so I owe ya, Rob.  ;-)

**Comment by skurfer** [comment]
@mrstanwell You’re welcome to try. Here’s a copy of the app that’s been built, but not signed. Our goal is to be able to run `spctl -av Quicksilver.app` and get no errors.

[Quicksilver.zip](https://github.com/quicksilver/Quicksilver/files/5415664/Quicksilver.zip)

We used to be able to get there with

    codesign -s "Developer ID Application" --deep Quicksilver.app

Lately, I’ve been trying to do it the right way (“inside out”) with

    codesign -s "Developer ID Application" Quicksilver.app/Contents/Frameworks/*.framework
    codesign -s "Developer ID Application" Quicksilver.app/Contents/PlugIns/*.qsplugin
    codesign -s "Developer ID Application" Quicksilver.app/Contents/Resources/QSDroplet.app
    codesign -s "Developer ID Application" Quicksilver.app/Contents/Library/Automator/*.action
    codesign -s "Developer ID Application" Quicksilver.app/Contents/Resources/*.plist
    codesign -s "Developer ID Application" Quicksilver.app/Contents/Info.plist
    codesign -s "Developer ID Application" Quicksilver.app

Running `codesign --verify --deep --strict --verbose=4 Quicksilver.app` will show that everything was signed correctly. If you check the app with Patrick Wardle’s [What’s Your Sign?][1] or Apple’s [`check-signature` tool][2], they both agree with `codesign` that the app is valid, but `spctl` will show the truth: **rejected**.

So, any ideas you have are welcome.

[1]: https://objective-see.com/products/whatsyoursign.html
[2]: https://developer.apple.com/library/archive/technotes/tn2206/_index.html#//apple_ref/doc/uid/DTS40007919-CH1-TNTAG211

**Comment by mrstanwell** [comment]
Hm.  Well, `spctl` will reject if there's anything wrong with the notarization pedigree.  I notice that the QS app dmg currently available on the site isn't notarized:
```
Downloads$ spctl -a -vv -t install  Quicksilver\ 1.6.1.dmg
Quicksilver 1.6.1.dmg: rejected
source=Unnotarized Developer ID
origin=Developer ID Application: Rob McBroom (ATVCND8EAP)
```

So, your signing is fine but perhaps something's going wrong with notarization...?  I have no idea if lack of valid notarization is why the API you need is blocking you, but in any event valid notarization is a must-have for any non-appstore apps on Catalina.  It's a bit of a headache for a CI pipeline, lemmetellya.  I recently ran across [a really helpful blog](https://scriptingosx.com/2019/09/notarize-a-command-line-tool/) from a fellow named Armin (@scriptingosx).  He's [got a script](https://gist.github.com/scriptingosx/9d72cf4a7a88df66c1c647b83825b26a#file-pkgandnotarize-sh) that is very similar to what my automation does to notarize, although I'm not building an installer and am instead notarizing the DMG with the app in (the same way QS distributes).

**Comment by skurfer** [comment]
I’ve never looked into or attempted notarization. I was hoping to avoid it. Maybe there’s no choice now.

**Comment by mrstanwell** [comment]
Yeah, at this point on Catalina any non-notarized app gets a scare popup that Apple can't verify it's not malicious software.  For now, you can still force the app to launch with admin privs... but who knows how long that's going to last.  It sounds like from your comments that they may be blocking API access to non-notarized apps as well.  Fortunately, notarizing isn't too difficult.  You do have to build your app as a "hardened runtime", and request any security entitlements you might need.  We didn't find this to be very difficult, and we have a pretty complicated app.  In my experience, apps are consistently notarized by Apple in under 5 min or so.  It stinks hugely to put a megacorp directly into the CI/CD pipeline, but... well... submission to our feudal overlords is non-optional.

[Armin's blog post](https://scriptingosx.com/2019/09/notarize-a-command-line-tool/) covers the bases, for the most part.  If there's something I can help with, or if you get stuck or run into questions, let me know and I'll do what I can.

**Comment by Calorion** [comment]
Is there a working Music plugin yet? I’m just trying to use a shortcut to change Music volume, but QuickSilver currently can’t do it.

**Comment by skurfer** [comment]
@Calorion No, but in the meantime, you could create your own triggers for volume. Something like `tell application "Music" …` ⇥ Run as AppleScript

**Comment by n8henrie** [comment]
FWIW I looked into making a new Music plugin a few weeks ago and the API just isn't available on Ventura. There is supposedly a beta API on Sonoma, so I'll look into it again soon.

**Comment by lgarron** [comment]
I gave this a poke and was able to set up a plugin with code that *should* be able to get songs from MusicKit.

Some funkiness:

- MusicKit requires using Swift. No issue, thanks to: https://github.com/quicksilver/plugin_template
- MusicKit requires targeting at least macOS 14.0. I'm able to set this value in `Common.xcconfig` but it gets overwritten every time I recompile Quicksilver itself. 😬
- The code is crashing due to:

```
This app has crashed because it attempted to access privacy-sensitive data without a usage description.  The app's Info.plist must contain an NSAppleMusicUsageDescription key with a string value explaining to the user how the app uses this data.
```

Unfortunately, this is still happening after I add a `NSAppleMusicUsageDescription` to the plugin's `Info.plist` as well `QSCore-Info.plist`, `QSFoundation-Info.plist`, and `QSInterface-Info.plist` in Quicksilver itself.

I'm sort of at an impasse here, since as far as I can tell I'm doing what the error tells me to do. (And this happens after deleting `/tmp/QS` and recompiling everything from scratch.)

**Comment by n8henrie** [comment]
> MusicKit requires using Swift. No issue, thanks to: https://github.com/quicksilver/plugin_template

Super psyched to see this getting put to use!
